## üí≤ Running code

In the following `.md` files, the dollar sign symbol within code chunks denotes lines of code that you should run in your terminal window, e.g.
```bash
$ echo "down here you can see the expected output of your commands"

down here you can see the expected output of your commands
```

## üèÅ 0. Clone repo to start exercises
Use the `git` command to download repo
```bash
$ git clone https://github.com/franciscozorrilla/SymbNET.git

Cloning into 'SymbNET'...
remote: Enumerating objects: 6611, done.
remote: Counting objects: 100% (6601/6601), done.
remote: Compressing objects: 100% (4397/4397), done.
remote: Total 6611 (delta 2525), reused 6107 (delta 2193), pack-reused 10
Receiving objects: 100% (6611/6611), 76.54 MiB | 2.96 MiB/s, done.
Resolving deltas: 100% (2525/2525), done.
Updating files: 100% (6770/6770), done.
```
Move into repo folder using `cd`
```bash
$ cd SymbNET
```

List contents of cloned repo using `ls`
```bash
$ ls

LICENSE			README.md		Snakefile		cluster_config.json	data			ensembles		genomes			models			plots			scripts			simulations
```

## üóª Contents

 - `/genomes/` ORF annotated protein fasta files (.faa)
 - `/models/` pre-generated CarveMe GEMs for simulation (.xml)
 - `/ensembles/` pre-generated ensemble models for network uncertainty (.xml) 
 - `/simulations/` pre-computed SMETANA simulations (.tsv)
 - `/data/` various metadata, taxonomic assignments, media files, etc.
 - `/scripts/` markdown, r-markdown, and jupyter notebooks for each exercise
 - `/plots/` visualization of results

## ‚õèÔ∏è Datasets

The following table describes in detail the 6 small bacterial communities of 5 species that we will consider for metabolic modeling. These include MAG,SAG, and reference genome-based communities; the samples also span the human gut, kefir, and soil habitats.

| Microbiome |Genome type| Condition | Species | Links |
| :---        | :--- | :---           | :---          |  :---         | 
| Human gut  | MAGs |Normal Glucose Tolerance (NGT, ERR260255) |  <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>   | <ul><li>[Paper](https://www.nature.com/articles/nature12198)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00001287#overview)</li></ul> |
| Human gut   | MAGs |Impaired Glucose Tolerance (IGT, ERR260172)      | <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>   |  <ul><li>[Paper](https://www.nature.com/articles/nature12198)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00001094?version=1.0#overview)</li></ul> |
| Human gut  | MAGs | Type II Diabetes (T2D, ERR260140)  | <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>  | <ul><li>[Paper](https://www.nature.com/articles/nature12198)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00001062?version=1.0#overview)</li></ul> |
| Human gut| Reference genomes | Reference genomes taken from RefSeq   | <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>  | <ul><li>[NCBI](https://www.ncbi.nlm.nih.gov/)</li></ul>   |
| Kefir | SAGs | Fermented with German grains (GER6)    | <ul><li>*L. mesenteroides*</li><li>*L. lactis*</li><li>*A. fabarum*</li><li>*L. kefiranofaciens*</li><li>*L. kefiri*</li></ul>  | <ul><li>[Paper](https://www.nature.com/articles/s41564-020-00816-5)</li></ul>   |
| Soil | MAGs |  Calcarosols from Uluru, Australia (ERR671933)   | <ul><li>f_Thermoleophilaceae</li><li>f_Herpetosiphonaceae</li><li>f_Phormidiaceae</li><li>f_Geodermatophilaceae</li><li>f_Rubrobacteraceae</li></ul>   | <ul><li>[Paper](https://academic.oup.com/gigascience/article/5/1/s13742-016-0126-5/2720982)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/studies/MGYS00000434)</li></ul>  |

## ü§ñ Cluster submission (optional)

The template `Snakefile` and config `cluster_config.json` can be modified by those wishing to experiment with higher throughput analysis.

## Move on to [exercise 1](https://github.com/franciscozorrilla/SymbNET/blob/main/scripts/1.carve_models.md)
