## üèÅ 0. Clone repo to start exercises

Use the `git` command to download repo
```bash
$ git clone https://github.com/franciscozorrilla/SymbNET.git
Cloning into 'SymbNET'...
remote: Enumerating objects: 6611, done.
remote: Counting objects: 100% (6601/6601), done.
remote: Compressing objects: 100% (4397/4397), done.
remote: Total 6611 (delta 2525), reused 6107 (delta 2193), pack-reused 10
Receiving objects: 100% (6611/6611), 76.54 MiB | 2.96 MiB/s, done.
Resolving deltas: 100% (2525/2525), done.
Updating files: 100% (6770/6770), done.
```

## üí≤ Running code

In the following `.md` files, the dollar sign symbol within code chunks denotes lines of code that you should run in your terminal window, e.g.
```bash
$ echo "down here you can see the expected output of your commands"
down here you can see the expected output of your commands
```

Move into repo folder using `cd`
```bash
$ cd SymbNET
```

Set the variable `ROOT` as the root of your `SymbNET` repo 
```bash
$ ROOT=$(pwd)
```

To navigate to the root folder from your terminal from any directory
```bash
$ cd $ROOT
```

List contents of cloned repo using `ls`
```bash
$ ls $ROOT
LICENSE			README.md		Snakefile		cluster_config.json	data			ensembles		genomes			models			plots			scripts			simulations
```

## üóª Contents

 - `/genomes/` open reading frame (ORF) annotated protein fasta files (.faa)
 - `/models/` pre-generated CarveMe GEMs for simulation (.xml)
 - `/ensembles/` pre-generated ensemble models for network uncertainty (.xml) 
 - `/simulations/` pre-computed SMETANA simulations (.tsv)
 - `/data/` various metadata, taxonomic assignments, media files, etc.
 - `/scripts/` markdown, r-markdown, and jupyter notebooks for each exercise
 - `/plots/` visualization of results

## ‚õèÔ∏è Datasets

The following table describes in detail the 6 small bacterial communities of 5 species that we will consider for metabolic modeling. These include MAG,SAG, and reference genome-based communities; the samples also span the human gut, kefir, and soil habitats.

| Microbiome |Genome type| Condition | Species | Links |
| :---        | :--- | :---           | :---          |  :---         | 
| 1. Human gut  | MAGs |Normal Glucose Tolerance (NGT, ERR260255) |  <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>   | <ul><li>[Paper](https://www.nature.com/articles/nature12198)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00001287#overview)</li></ul> |
| 2. Human gut   | MAGs |Impaired Glucose Tolerance (IGT, ERR260172)      | <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>   |  <ul><li>[Paper](https://www.nature.com/articles/nature12198)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00001094?version=1.0#overview)</li></ul> |
| 3. Human gut  | MAGs | Type II Diabetes (T2D, ERR260140)  | <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>  | <ul><li>[Paper](https://www.nature.com/articles/nature12198)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00001062?version=1.0#overview)</li></ul> |
| 4. Human gut| Reference genomes | Reference genomes taken from RefSeq   | <ul><li>*B. uniformis*</li><li>*R. bromii*</li><li>*B. wexlerae*</li><li>*E. rectale*</li><li>*F. saccharivorans*</li></ul>  | <ul><li>[NCBI](https://www.ncbi.nlm.nih.gov/)</li></ul>   |
| 5. Kefir | SAGs | Fermented with German grains (GER6)    | <ul><li>*L. mesenteroides*</li><li>*L. lactis*</li><li>*A. fabarum*</li><li>*L. kefiranofaciens*</li><li>*L. kefiri*</li></ul>  | <ul><li>[Paper](https://www.nature.com/articles/s41564-020-00816-5)</li></ul>   |
| 6. Soil | MAGs |  Calcarosols from Uluru, Australia (ERR671933)   | <ul><li>f_Thermoleophilaceae</li><li>f_Herpetosiphonaceae</li><li>f_Phormidiaceae</li><li>f_Geodermatophilaceae</li><li>f_Rubrobacteraceae</li></ul>   | <ul><li>[Paper](https://academic.oup.com/gigascience/article/5/1/s13742-016-0126-5/2720982)</li><li>[metaGEM](https://github.com/franciscozorrilla/metaGEM_paper)</li><li>[MGnify](https://www.ebi.ac.uk/metagenomics/studies/MGYS00000434)</li></ul>  |

As a group, choose one of the 6 bacterial communities: `gut_normal`, `gut_impaired`, `gut_t2d`, `gut_refseq`, `kefir`, or `soil`.

Use the `COMM` variable to select your community, e.g.

```bash
$ COMM=gut_normal
```

Move to folder containing ORF-annotated protein fasta files (.faa) and use the `ls` command to list files in the folder, e.g.
```bash
$ cd genomes/$COMM
$ ls
ERR260255_bin.14.p.faa	ERR260255_bin.19.p.faa	ERR260255_bin.24.s.faa	ERR260255_bin.7.p.faa	ERR260255_bin.9.s.faa
```

Inspect genomes to make sure that they contain ORF-annotated protein sequences
```bash
$ head -n 25 ERR260255_bin.14.p.faa
>NODE_1_length_135324_cov_8.930368_1 # 150 # 2264 # -1 # ID=1_1;partial=00;start_type=ATG;rbs_motif=AGGA;rbs_spacer=5-10bp;gc_cont=0.502
MGEYVNPAKIFGENVFNDTVMQERLPKNVYKKLKKIIEEGGELDLATADTIAHEMKEWAI
EKGATHYTHWFQPLTGTTAEKHDSFITAPMPNGKVLMSFSGKELIKGEPDASSFPSGGLR
ATFEARGYTAWDCTSPAFVRQDAAGATLCIPTAFCSYTGEALDQKTPLLRSMEALNVQSL
RLLKLFGNTTSKKVTPSVGPEQEYFLVDAEKFLQRKDLIYTGRTLFGAMPPKGQEMDDHY
FGTIRQRIASFMKEVNIELWKMGVTAKTQHNEVAPAQHELASIYSEANVAVDNNQLVMQT
LKRVACQHGLKCLLHEKPFAGVNGSGKHNNWSITTDDGINLLEPGKTPHENIQFLLVLSC
VLKAVDVHADLLRESAADPGNDHRLGANEAPPAIISVFLGEQLEDVVEQLISTGNATKSK
KEGVLETGVKTLPDLKKDATDRNRTSPFAFTGNKFEFRMVGSRDSVAAPNIVLNTIVAEA
FRDACDVLEGAENFEDAVHDLIKKNLSEHQRIIFNGDGYADEWLAEAERRGLPNIKSMVY
AIPALTTDTAIKLFGDFKVFTEAELVSRAEVKFENYAKTINIEAKTMIDMASKQIIPAVI
KYATSLAGSINTITAAGVTAVGVQKNLLNETSALLEETQKALDELIAIENAGCEMEDGEA
KAKYYYEKVAPAMEALRAPVDKLEMIVDKEMWPMPSYGDLMFEV*
>NODE_1_length_135324_cov_8.930368_2 # 2679 # 3542 # -1 # ID=1_2;partial=00;start_type=ATG;rbs_motif=AGGAGG;rbs_spacer=11-12bp;gc_cont=0.469
MLVSAKEMLDKAKAGHYAVGQFNINNLEWTKSILLTAQEMNSPVILGVSEGAGKYMTGFK
TVAAMVKAMIEELNITVPVALHLDHGTYEGCYKCIEAGFSSIMFDGSHYPIEENVAKTKE
LVAVCKEKGMSLEAEVGSIGGEEDGVIGMGECADPNECKMIADLGVDMLAAGIGNIHGVY
PANWQGLSFETLAAVQELTGTMPLVLHGGTGIPDDMIKKAIDLGVSKINVNTECQLAFAA
ATRKYIEEGKDQQGKGFDPRKLLAPGAEAIKATVREKIELFGSANKA*
>NODE_1_length_135324_cov_8.930368_3 # 3725 # 4624 # -1 # ID=1_3;partial=00;start_type=ATG;rbs_motif=AGGAG;rbs_spacer=5-10bp;gc_cont=0.496
MESKKKMLFIFNPFSGKAQIKSKLFEIIDVFVKGGYEVIVHPTQAVGDGFEKTKELAPQV
DLVVCSGGDGTLDEVVSGLMEVDQRVPIGYIPAGSTNDFANSLSISKDMVQAAKDIIEGN
LYSCDVGAFNNDSFVYIAAFGLFTDVSYETDQHMKNILGHLAYLLEGSKRIWNVPTYWIK
VEANGETFEGEYIYGMVTNAKSVGGFKNLPGQDVRLDDGLFEVTLIKRPKNPLELNEIIA
SLLMQEDHTDLIDSFKTDRIIIESNQEISWTLDGEYGGDHGYVCIEARKHAMELVLNNK*
```

## ‚öôÔ∏è Translating MAGs: from DNA to protein sequences

If your genomes are in the form of DNA fasta files, run prodigal to translate to protein and annotate open reading frames (ORFs).
Download a *Eubacterium rectale* genome from RefSeq using `wget`

```bash
$ wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/209/935/GCF_000209935.1_ASM20993v1/GCF_000209935.1_ASM20993v1_genomic.fna.gz
```
Use `gunzip` to unzip the file
```bash
$ gunzip GCF_000209935.1_ASM20993v1_genomic.fna.gz
```
Use `head` to view and verify the contents of the file
```bash
$ head GCF_000209935.1_ASM20993v1_genomic.fna
>NC_021010.1 Eubacterium rectale DSM 17629 draft genome
TGAATGAACGAATTAGCCAAGTTTAACTTGACTGAGATCAACTTTTAACATGAGAGTTTGATCCTGGCTCAGGATGAACG
CTGGCGGCGTGCTTAACACATGCAAGTCGAACGAAGCACTTTATTTGATTTCCTTCGGGACTGATTATTTTGTGACTGAG
TGGCGGACGGGTGAGTAACGCGTGGGTAACCTGCCTTGTACAGGGGGATAACAGTTGGAAACGGCTGCTAATACCGCATA
AGCGCACAGCATCGCATGATGCAGTGTGAAAAACTCCGGTGGTATAAGATGGACCCGCGTTGGATTAGCTAGTTGGTGAG
GTAACGGCCCACCAAGGCGACGATCCATAGCCGACCTGAGAGGGTGACCGGCCACATTGGGACTGAGACACGGCCCAAAC
TCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGCGAAGAAGTAT
TTCGGTATGTAAAGCTCTATCAGCAGGGAAGATAATGACGGTACCTGACTAAGAAGCACCGGCTAAATACGTGCCAGCAG
CCGCGGTAATACGTATGGTGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGCAGGCGGTGCGGCAAGTCTGATG
TGAAAGCCCGGGGCTCAACCCCGGTACTGCATTGGAAACTGTCGTACTAGAGTGTCGGAGGGGTAAGCGGAATTCCTAGT
```

Run prodigal to generate ORF annotated protein fasta file desired for CarveMe input
```bash
$ prodigal -i GCF_000209935.1_ASM20993v1_genomic.fna -a proteins.faa
```

Use `tail` to verify contents of prodigal output

```bash
$ tail proteins.faa
>NC_021010.1_3164 # 3343617 # 3344390 # 1 # ID=1_3164;partial=00;start_type=ATG;rbs_motif=AGGAGG;rbs_spacer=5-10bp;gc_cont=0.386
MSNAINVNYMTRAYNQYQQKNAAKDQEKEDTRFADSVREKSEASGSIVGNSKIGSVSAKD
MTMVEYKQYIYNKISQIPMHPTRAGESISVTISEAGFEAMKNDPEYEAWVLNDLQVGWSQ
PDKWSGICGGAFSTIYYGASKEECHAEMWSAGYNNGNGGKIFNDKSKNSFWERRIENKKR
IENQVKKQQEKKRIQKKQAERTAYEEYAQNKRLSAQDARTRLVAESSSSVKTVTISQAVA
SYEANFTMAVSVSNKTQ*
>NC_021010.1_3165 # 3344436 # 3344951 # 1 # ID=1_3165;partial=01;start_type=ATG;rbs_motif=AGGAG;rbs_spacer=5-10bp;gc_cont=0.341
MGINGVSSYANTYTYGNVKNENVKPSKNSEKNYYAGYTMNSANKFKNVSDYSKYLTNKYK
CLTPCKNASVLIDGSVMRKACGDEKTAKWLEENLAIMPDVIRNAQKAAISHGSKLISVEF
KFTNNGTEMTTCGIFGETGTDSEIDKWLERMKEDKEKEDKKTENMIAIEATT
```

As an optional exercise, try running prodigal on one of the DNA fasta file genomes generated on the previous days.

## üôã Discussion questions

 - Do you expect certain microbial communities to be cooperative or competitve? Why?
 - Rank the communities in the table above from most cooperative to most competitve and justify your answer.

## Move on to [exercise 1](https://github.com/franciscozorrilla/SymbNET/blob/main/scripts/1.carve_models.md)
