## üèÅ Exercise 0: The data

 - Do you expect certain microbial communities to be cooperative or competitve? Why?
 - Rank the communities in the table above from most cooperative to most competitve and justify your answer.

## ü™ö Exercise 1: CarveMe model reconstruction

 - How does CarveMe generate metabolic models? 
 - How does ORF-annotation of genome DNA influence model reconstruction? What are strenghts and weaknesses of community standard/new tools?
    * [Prodigal](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-119)
    * [Prokka](https://academic.oup.com/bioinformatics/article/30/14/2068/2390517)
    * [smORF](https://www.sciencedirect.com/science/article/pii/S1931312820306193?via%3Dihub)
 - What are the pros and cons of relying on the BiGG database for model reconstruction?
 - How does the choice of gap-filling media affect model reconstruction?
 - Why are there specialized model templates?

## üßÆ Exercise 2: CarveMe model visualization

 - Approximately how many genes do you expect to find in a typical prokaryotic genome? What percentage do you expect to be metabolism encoding?
 - Are the number of metabolites, reactions, and genes in your community's models comparable to other metabolic reconstructions?
 - Is there variation in the metabolism of conspecific models reconstructed from different metagenomes?

## üï∏Ô∏è Exercise 3: SMETANA detailed algorithm

 - How does the SMETANA detailed algorithm work at a high level?
 - What are the underlying assumptions?
 - What does each metric measure and how is it calculated?
    * SCS
    * MUS
    * MPS
    * SMETANA
 - What does a SMETANA score of 0 mean? What does a SMETANA score of 1 mean?
 - Why do we run multiple simulations and take averages of SMETANA scores? Why is it important to use the `--zeros` flag in this case?
 - How does choice of simulation media affect the SMETANA detailed algorithm output?
 - What does the `--molweight` flag do? How can it be useful?

## üìà Exercise 4: SMETANA detailed visualization

-   Do the predictions make sense from a biological perspective?
-   Look at the receiver species on your community‚Äôs alluvial diagram, are they known to be auxotrophic for the metabolites we predicted them to uptake?
-   Look at the donor species on your community‚Äôs alluvial diagram, are they known to be prototrophic for the metabolites we predicted them to donate?
-   Is there supporting literature for the predicted interactions in your community?
-   Can this approach be used for hypothesis generation in microbial ecology?

## üåê Exercise 5: SMETANA global algorithm

 - How does the SMETANA global algorithm work?
 - What are the underlying assumptions?
 - What does each metric measure and how is it calculated?
   * MIP
   * MRO
 - Does choice of media affect global score calculation?

## ‚öñÔ∏è Exercise 6: SMETANA global visualization

 - How did your predicted competitiveness/cooperativeness rankings compare with the results from the plot above? Are you surprised by the results?
 - Why is there variation within communities, given that simulations were carried out using the same models?
 - Why might the simulations generated by some of our communities fall outside of the original publication's cooperation-coompetition plot (community size N=5)?

## üé∞ Exercise 7: CarveMe ensemble models (optional)

## ‚ùì Exercise 8: CarveMe network uncertainity (optional)

## üí¨ General Discussion

- What assumptions are made in genome-scale modelling approach
- What input data is required?
- What type of data can be integrated, how, and what for? E.g. transcriptomics, proteomics, metabolomics
- How would you test the model predictions experimentally?
- Can the model predict unexpected interactions?
- Do you think this approach could be used in your project?  
